File: pnix-clj/clj-meta/src/pnix/clj_meta/io.clj
(ns pnix.clj-meta.io
  "Pnix-agnostic, capability-gated read-only host I/O substrate.

  This namespace knows nothing about pnix effect requests or values. Active
  runtimes adapt these four operations at their own interop boundary."
  (:import [java.nio.charset CharacterCodingException StandardCharsets]
           [java.nio.file Files LinkOption NoSuchFileException
            NotDirectoryException Path Paths]))

(def file-read-capability "file-read")

(defn- capability-name [cap]
  (cond
    (keyword? cap) (name cap)
    (string? cap) cap
    :else (str cap)))

(defn file-read-granted? [granted]
  (boolean (some #(= file-read-capability (capability-name %)) granted)))

(defn- require-file-read! [granted]
  (when-not (file-read-granted? granted)
    (throw (ex-info "file-read capability denied"
                    {:error-class :capability-denied
                     :capability file-read-capability}))))

(defn- as-path ^Path [path]
  (Paths/get (str path) (make-array String 0)))

(defn- classify-path [^Path path]
  (cond
    (Files/isSymbolicLink path) "symlink"
    (Files/isDirectory path (make-array LinkOption 0)) "directory"
    (Files/isRegularFile path (make-array LinkOption 0)) "regular"
    :else "unknown"))

(defn path-exists [path granted]
  (require-file-read! granted)
  (Files/exists (as-path path) (make-array LinkOption 0)))

(defn file-type [path granted]
  (require-file-read! granted)
  (let [p (as-path path)]
    (when-not (or (Files/exists p (make-array LinkOption 0))
                  (Files/isSymbolicLink p))
      (throw (ex-info "file type target not found"
                      {:error-class :not-found})))
    (classify-path p)))

(defn read-utf8 [path granted]
  (require-file-read! granted)
  (try
    (Files/readString (as-path path) StandardCharsets/UTF_8)
    (catch NoSuchFileException e
      (throw (ex-info "read target not found" {:error-class :not-found} e)))
    (catch CharacterCodingException e
      (throw (ex-info "read target is not UTF-8" {:error-class :invalid-utf8} e)))
    (catch java.io.IOException e
      (throw (ex-info "read failed" {:error-class :io-error} e)))))

(defn read-dir [path granted]
  (require-file-read! granted)
  (try
    (with-open [stream (Files/newDirectoryStream (as-path path))]
      (into (sorted-map)
            (map (fn [^Path entry]
                   [(str (.getFileName entry)) (classify-path entry)]))
            (iterator-seq (.iterator stream))))
    (catch NoSuchFileException e
      (throw (ex-info "directory not found" {:error-class :not-found} e)))
    (catch NotDirectoryException e
      (throw (ex-info "target is not a directory" {:error-class :not-directory} e)))
    (catch java.io.IOException e
      (throw (ex-info "directory read failed" {:error-class :io-error} e)))))

(defn report []
  (let [granted #{file-read-capability}
        denied (try
                 (path-exists "deps.edn" #{})
                 false
                 (catch clojure.lang.ExceptionInfo e
                   (= :capability-denied (:error-class (ex-data e)))))]
    {:schema :clj-meta.io.v1
     :ready (and denied
                 (path-exists "deps.edn" granted)
                 (= "regular" (file-type "deps.edn" granted))
                 (contains? (read-dir "." granted) "src"))
     :capability file-read-capability
     :effects ["path-exists" "open" "file-type" "read-dir"]}))

(defn -main [& _]
  (let [result (report)]
    (prn result)
    (when-not (:ready result)
      (System/exit 1))))

File: pnix-rs/rs-meta/src/io.rs
//! Pnix-agnostic, capability-gated read-only host I/O substrate.
//!
//! This file is source-shared by pnix-rs's interop boundary. It deliberately
//! contains no pnix request/value semantics.

use std::path::Path;

pub const FILE_READ_CAPABILITY: &str = "file-read";

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct MetaIoError {
    pub error_class: String,
    pub message: String,
}

fn error(error_class: &str, message: &str) -> MetaIoError {
    MetaIoError {
        error_class: String::from(error_class),
        message: String::from(message),
    }
}

fn require_file_read(granted: &[String]) -> Result<(), MetaIoError> {
    if granted.iter().any(|item| item == FILE_READ_CAPABILITY) {
        Ok(())
    } else {
        Err(error("capability-denied", "file-read capability denied"))
    }
}

fn classify(path: &Path) -> Result<String, MetaIoError> {
    let metadata = std::fs::symlink_metadata(path)
        .map_err(|_| error("not-found", "file type target not found"))?;
    let kind = metadata.file_type();
    if kind.is_symlink() {
        Ok(String::from("symlink"))
    } else if kind.is_dir() {
        Ok(String::from("directory"))
    } else if kind.is_file() {
        Ok(String::from("regular"))
    } else {
        Ok(String::from("unknown"))
    }
}

pub fn path_exists(path: &str, granted: &[String]) -> Result<bool, MetaIoError> {
    require_file_read(granted)?;
    Ok(Path::new(path).exists())
}

pub fn file_type(path: &str, granted: &[String]) -> Result<String, MetaIoError> {
    require_file_read(granted)?;
    classify(Path::new(path))
}

pub fn read_utf8(path: &str, granted: &[String]) -> Result<String, MetaIoError> {
    require_file_read(granted)?;
    let bytes = std::fs::read(path).map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            error("not-found", "read target not found")
        } else {
            error("io-error", "read failed")
        }
    })?;
    String::from_utf8(bytes).map_err(|_| error("invalid-utf8", "read target is not UTF-8"))
}

pub fn read_dir(path: &str, granted: &[String]) -> Result<Vec<(String, String)>, MetaIoError> {
    require_file_read(granted)?;
    let entries = std::fs::read_dir(path).map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            error("not-found", "directory not found")
        } else if e.kind() == std::io::ErrorKind::NotADirectory {
            error("not-directory", "target is not a directory")
        } else {
            error("io-error", "directory read failed")
        }
    })?;
    let mut out = Vec::new();
    for item in entries {
        let entry = item.map_err(|_| error("io-error", "directory read failed"))?;
        let name = entry.file_name().to_string_lossy().into_owned();
        let kind = classify(&entry.path())?;
        out.push((name, kind));
    }
    out.sort();
    Ok(out)
}

#[allow(dead_code)]
pub fn io_check() -> bool {
    println!("[io-check (capability-gated read-only host I/O)]");
    let none: Vec<String> = Vec::new();
    let granted = vec![String::from(FILE_READ_CAPABILITY)];
    let denied = matches!(
        path_exists("Cargo.toml", &none),
        Err(MetaIoError { ref error_class, .. }) if error_class == "capability-denied"
    );
    let ready = denied
        && path_exists("Cargo.toml", &granted) == Ok(true)
        && file_type("Cargo.toml", &granted) == Ok(String::from("regular"))
        && read_utf8("Cargo.toml", &granted)
            .map(|text| text.contains("name = \"rs-meta\""))
            == Ok(true)
        && read_dir("src", &granted)
            .map(|entries| entries.iter().any(|(name, kind)| name == "main.rs" && kind == "regular"))
            == Ok(true);
    if ready {
        println!("  => PASS (file-read denied by default; four read effects ready)");
    } else {
        println!("  => FAIL");
    }
    ready
}
